src/HttpServer.php

```diff
@@ -104,42 +104,37 @@ public function handleConnection(ConnectionInterface $connection)
     /** @internal */
     public function handleBody(RequestInterface $request, ConnectionInterface $connection)
     {
-        $bodyBuffer = '';
-        $that = $this;
-
         $protection = new CloseProtectionStream($connection);
-
         if ($this->isChunkedEncodingActive($request)) {
+            // Add ChunkedDecoder to stream
             $chunkedDecoder = new ChunkedDecoder($protection);
-            $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $connection, $that) {
-                $bodyBuffer .= $chunk;
-                if (strlen($chunk) == 0) {
-                    $request = $request->withBody(RingCentral\Psr7\stream_for($bodyBuffer));
-                    $that->handleRequest($connection, $request);
-                }
-            });
+            $bodyStream = new HttpBodyStream($chunkedDecoder);
+            $request = $request->withBody($bodyStream);
+            $this->handleRequest($connection, $request);
             return;
         }
 
-        $contentLengthArray = $request->getHeader('Content-Length');
-
-        if (!$request->hasHeader('Content-Length') || $contentLengthArray[0] == 0) {
+        if (!$request->hasHeader('Content-Length')) {
+            // Request hasn't defined 'Content-Length' will ignore rest of the request
+            // and ends the stream
+            $bodyStream = new HttpBodyStream($protection);
+            $request = $request->withBody($bodyStream);
             $this->handleRequest($connection, $request);
+            $bodyStream->close();
```

Comment:
Does it make sense to close the stream *before* or *after* it is passed to the callback function?

A closed stream will not emit an `end` event, however, it's state could be checked via `isReadable()`.

I'm undecided, what do you think?